src/components/Shell/RoomShareDialog.tsx

```diff
@@ -0,0 +1,188 @@
+import {
+  Alert,
+  AlertColor,
+  Button,
+  Checkbox,
+  Dialog,
+  DialogActions,
+  DialogContent,
+  DialogContentText,
+  DialogTitle,
+  IconButton,
+  FormControlLabel,
+  TextField,
+  Tooltip,
+} from '@mui/material'
+import CloseIcon from '@mui/icons-material/Close'
+
+import { AlertOptions } from 'models/shell'
+import { useEffect, useState, SyntheticEvent } from 'react'
+import { encodePassword, sleep } from 'utils'
+
+export interface RoomShareDialogProps {
+  isOpen: boolean
+  handleClose: () => void
+  roomId: string
+  password: string
+  showAlert: (message: string, options?: AlertOptions) => void
+  copyToClipboard: (
+    content: string,
+    alert: string,
+    severity: AlertColor
+  ) => Promise<void>
+}
+
+export function RoomShareDialog(props: RoomShareDialogProps) {
+  const [isAdvanced, setIsAdvanced] = useState(false)
+  const [isUnderstood, setIsUnderstood] = useState(false)
+  const [password, setPassword] = useState('')
+  const [passThrottled, setPassThrottled] = useState(false)
+  const handleClose = () => {
+    props.handleClose()
+    setPassword('')
+  }
+
+  useEffect(() => {
+    if (!isAdvanced) setIsUnderstood(false)
+  }, [isAdvanced])
+
+  useEffect(() => {
+    if (!isUnderstood) setPassword('')
+  }, [isUnderstood])
+
+  const url = window.location.href.split('#')[0]
+
+  const copyWithPass = async () => {
+    const encoded = await encodePassword(props.roomId, password)
+    if (encoded === props.password) {
+      const params = new URLSearchParams()
+      params.set('secret', props.password)
+      await props.copyToClipboard(
+        `${url}#${params}`,
+        'Private room URL with password copied to clipboard',
+        'warning'
+      )
+      handleClose()
+    } else {
+      setPassThrottled(true)
+      props.showAlert('Incorrect password entered. Please wait 2s to retry.', {
+        severity: 'error',
+      })
+      await sleep(2000)
```

Comment:
I really like that we're throttling password retries to prevent brute-forcing the password. Great idea!